### DISCUSSION TITLE
Should HTMX use getElementById or querySelector instead of querySelectorAll when replacing based on id?

### DISCUSSION THREAD

**Discussion by CaliViking** [bigskysoftware/htmx#2472] [Ideas]
I may be on thin ice here, as I am not sure that there is an actual performance difference. However, I have noticed that HTMX becomes slower when there are many html elements in the DOM.

In the function oobSwap(oobValue, oobElement, settleInfo) https://github.com/bigskysoftware/htmx/blob/8e26d12c33caf7dc99c49d4918e7b8abd8028f98/src/htmx.js#L817 we specifically look for an id, but we are using querySelectorAll instead of getElementById or querySelector.

According to https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id - every id in the document should be unique, hence there should be no reason to use querySelectorAll, as only one element can be returned.

Would it be faster to use getElementById or querySelector than querySelectorAll? There seem to be different opinions on this on the internet, but it seems logical that getElementById should be faster as it is more specific. It should be a simple lookup.

Is there a reason that the code is using querySelectorAll today? It seems to be in conflict with the HTML specification.

Thanks in advance,
Niels

**Comment by Telroshan** [comment]
Hey, according to [hx-swap-oob's documentation](https://htmx.org/attributes/hx-swap-oob/)
> If a selector is given, all elements matched by that selector will be swapped. If not, the element with an ID matching the new content will be swapped.

If you look at this line in `oobSwap` 
https://github.com/bigskysoftware/htmx/blob/8e26d12c33caf7dc99c49d4918e7b8abd8028f98/src/htmx.js#L824
You can see that place where the selector is no longer necessarily a unique ID selector, but can be any valid CSS selector

As for the performance aspect, never trust the random opinions on the internet ; measure it, it's the only way to know for sure!
